File: api/src/main/java/de/storyteller/api/service/chapter/ChapterService.java
package de.storyteller.api.service.chapter;

import de.storyteller.api.v1.dto.chapter.AddChapterRequest;
import de.storyteller.api.v1.dto.chapter.ChapterDTO;
import de.storyteller.api.v1.dto.chapter.EditChapterRequest;

import java.util.List;
import java.util.Optional;

/**
 * Service for chapters
 */
public interface ChapterService {

    /**
     * Create a chapter
     * @param chapter the chapter to create
     * @return the created chapter
     */
    ChapterDTO createChapter(AddChapterRequest chapter);

    /**
     * Get a chapter by its id
     * @param chapterId the id of the chapter
     * @return the chapter with the given id
     */
    Optional<ChapterDTO> getChapterById(String chapterId);

    /**
     * Update a chapter
     * @param chapter the chapter to update
     * @return the updated chapter
     */
    ChapterDTO updateChapter(EditChapterRequest chapter);

    /**
     * Get all chapters
     * @return a list of all chapters
     */
    List<ChapterDTO> getAllChapters();

    /**
     * Get all chapters of a book
     * @param bookId the id of the book
     * @return a list of all chapters of the book
     */
    List<ChapterDTO> getChaptersByBookId(String bookId);

    /**
     * Get all published chapters of a book
     * @param bookId the id of the book
     * @return a list of all published chapters of the book
     */
    List<ChapterDTO> getPublishedChaptersByBookId(String bookId);

    /**
     * Get the first n published chapters of a book
     * @param bookId the id of the book
     * @param count the number of chapters to get
     * @return a list of the first n published chapters of the book
     */
    List<ChapterDTO> getFirstNPublishedChapters(String bookId, int count);
}

File: webapp/storyteller/src/app/service/chapter.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {environment} from "../../environments/environment";
import {AddChapterRequest, Chapter, EditChapterRequest} from "../model/chapter";
import {Observable} from "rxjs";

@Injectable({
  providedIn: 'root'
})

/**
 * Service for Chapters
 */
export class ChapterService {

  constructor(private http: HttpClient) {
  }

  /**
   * Get chapter by id
   * @param chapterId the id of the chapter
   */
  getChapterById(chapterId: string): Observable<Chapter> {
    return this.http.get<Chapter>(`${environment.apiUrl}/chapters/${chapterId}`);
  }

  /**
   * Adds a new chapter
   * @param chapter the chapter to add
   */
  addChapter(chapter: AddChapterRequest): Observable<Chapter> {
    return this.http.post<Chapter>(`${environment.apiUrl}/chapters/add`, chapter);
  }

  /**
   * Updates a chapter
   * @param chapter the updated chapter
   */
  updateChapter(chapter: EditChapterRequest): Observable<Chapter> {
    return this.http.put<Chapter>(`${environment.apiUrl}/chapters/update`, chapter);
  }

  /**
   * Get all published chapters by their book id
   * @param bookId
   */
  getPublishedChaptersByBookId(bookId: string): Observable<Chapter[]> {
    return this.http.get<Chapter[]>(`${environment.apiUrl}/chapters/${bookId}/published`);
  }

  /**
   * Get the first n published chapters by their book id
   * @param bookId the id of the book to get the chapters from
   * @param n the number of chapters to get
   */
  getNPublishedChaptersByBookId(bookId: string, n: number): Observable<Chapter[]> {
    return this.http.get<Chapter[]>(`${environment.apiUrl}/chapters/${bookId}/published/${n}`);
  }
}
